Improve v3 class generating

dstepanov opened the pull request:
Introducing `AbstractBeanDefinition2`, `AbstractBeanDefinitionReference2` with improvements:
- passing most of the parameters in the constructor instead of overriding methods
- initializing method/field injection in the static block allowing Graal to perform build-time optimizations
- generating annotation metadata with shared annotation default variables, reducing significantly duplicate calls mostly for config classes
- improving how annotation metadata are generated
- use try-resource syntax for bean context push/pop
- `AbstractBeanDefinitionReference2` includes more precompilated variables at compile-time
- bean references are generated with prefix `$Reference` instead of the previous `Class` for better understanding what does the class represent
- generate all executable methods in one class prefixed "$Exec" instead of the previous one class per method
- changed to generate bean definition classes with prefix "$Definition" instead of previous "Definition" to better distinguish generated class
- use new map implementation `ImmutableSortedStringsArrayMap` for annotation classes instead of `Map.of`. The implementation is using a sorted `String` keys array for lookup and an extra array for keys. The size of the collection should be the same as in `Map.of`
- Refactored a lot of original `AbstractBeanDefinition`, most of the resolve code is now shared between field/constructor/methods methods
- Moved most of the code that decided how to resolve the argument to the compile-time
- Including `AbstractParametrizedBeanDefinition` main methods inside `AbstractBeanDefinition2` for the benefit of simplicity to have one super-class
- Resolving generic types, known qualifiers, and inner configuration types at the compile-time

dstepanov committed:
Improve bean definition class generation

dstepanov commented:
Not sure what to do with `containsProperties` I have discovered that it's not correctly implemented in the current version:
```
@UsedByGeneratedCode
    protected final boolean containsProperties(@SuppressWarnings("unused") BeanResolutionContext resolutionContext, BeanContext context, String subProperty) {
        boolean isSubProperty = StringUtils.isNotEmpty(subProperty);
// This hack is to skip the `if` if the configuration contains bean injects
        if (!isSubProperty && !requiredComponents.isEmpty()) {
            // if the bean requires dependency injection we disable this optimization
            return true;
        }
        if (isConfigurationProperties && context instanceof ApplicationContext) {
            AnnotationMetadata annotationMetadata = getAnnotationMetadata();
            ApplicationContext appCtx = (ApplicationContext) context;
// This is always true because it return empty array
            if (annotationMetadata.getValue(ConfigurationProperties.class, "cliPrefix").isPresent()) {
                return true;
            } else {
                String path = getConfigurationPropertiesPath(resolutionContext);
                return appCtx.containsProperties(path);
            }

        }
        return false;
    }
```
In the new version I have made `requiredComponents` lazy, so not sure what would be the best to implement this method, for now, it return the value if `isConfigureProperties`
The best solution would be to generate `if (containsProperties` only for the group of property injection but that requires collecting all of the injections and generating the method later.

graemerocher commented:
Thanks @dstepanov will review on Monday, can you provide some data on what the improvements are like for:

- Memory consumption (GraalVM & JIT)
- Startup (if any?)
- Image size for native

I think as well before we introduce a new metadata format we should address some problems with the old one, I will elaborate more on this in my review

graemerocher commented on inject/src/main/java/io/micronaut/context/AbstractBeanDefinition2.java:
```diff
@@ -0,0 +1,2338 @@
+        MethodReference constructorMethodRef = (MethodReference) constructor;
+        Argument<?> argument = resolveArgument(context, argIndex, constructorMethodRef.arguments);
+        final Class<?> beanType = argument.getType();
+        if (beanType == BeanResolutionContext.class) {
+            return resolutionContext;
+        } else if (argument.isArray()) {
+            Collection beansOfType = getBeansOfTypeForConstructorArgument(resolutionContext, context, argument);
+            return beansOfType.toArray((Object[]) Array.newInstance(beanType.getComponentType(), beansOfType.size()));
+        } else if (Collection.class.isAssignableFrom(beanType)) {
+            Collection beansOfType = getBeansOfTypeForConstructorArgument(resolutionContext, context, argument);
+            return coerceCollectionToCorrectType(beanType, beansOfType);
+        } else if (Stream.class.isAssignableFrom(beanType)) {
+            return streamOfTypeForConstructorArgument(resolutionContext, context, argument);
+        } else if (argument.isOptional()) {
+            return findBeanForConstructorArgument(resolutionContext, context, argument);
+        } else {
```
Can we precompute this at compilation time and just dispatch to the right method? All of this `isAssingableFrom` checks can be calculated at build time. What I would like to see if we are going to switch metadata format is more fine grained methods like `getBeanArrayForConstructorArgument` that are directly dispatched to

dstepanov replied:
It should be possible

graemerocher commented on inject/src/main/java/io/micronaut/context/AbstractBeanDefinition2.java:
```diff
@@ -0,0 +1,2338 @@
+            beanType = beanType.getSuperclass();
+        }
+        return false;
+    }
```
Can we precompute these at build time? We should know at build time whether an argument is an inner configuration and not require this logic for the `v2` format

graemerocher commented on inject/src/main/java/io/micronaut/context/AbstractBeanDefinition2.java:
```diff
@@ -0,0 +1,2338 @@
+            return argument.getFirstTypeVariable().orElse(null);
+        }
+        return genericType;
+    }
```
Can we precompute this build time? There should be no reason to resolve the generic type if we already know it at build time

graemerocher commented on inject/src/main/java/io/micronaut/context/AbstractBeanDefinition2.java:
```diff
@@ -0,0 +1,2338 @@
+            }
+            return qualifier;
+        }
+    }
```
I would like to see some of this logic resolved at build time. For the simple case it should not be necessary to have all of this logic and we should be able to compute the qualifier at build time and only fall back for the more complex logic if we know it is iterable (EachBean/EachProperty)

graemerocher commented on inject/src/main/java/io/micronaut/context/AbstractParametrizedBeanDefinition2.java:
```diff
@@ -0,0 +1,113 @@
+        return Arrays.stream(getConstructor().getArguments())
+            .filter(arg -> {
+                Optional<Class<? extends Annotation>> qualifierType = arg.getAnnotationMetadata().getAnnotationTypeByStereotype(AnnotationUtil.QUALIFIER);
+                return qualifierType.isPresent() && qualifierType.get() == Parameter.class;
```
Should be able to use `getAnnotationNameByStereotype` here instead which will be cheaper

graemerocher commented on inject/src/main/java/io/micronaut/inject/ExecutableMethodsDefinition.java:
```diff
@@ -0,0 +1,56 @@
+ * @author Denis Stepanov
+ * @since 3.0
+ */
+public interface ExecutableMethodsDefinition<T> {
```
Needs `@Nullable` / `@NonNull` on the method signatures also do we want this to be public API? Seems internal and should be `@Internal`?

graemerocher commented on inject/src/main/java/io/micronaut/inject/writer/BeanDefinitionWriter.java:
```diff
@@ -15,36 +15,9 @@
-import io.micronaut.core.annotation.Nullable;
+import io.micronaut.context.*;
+import io.micronaut.context.annotation.*;
+import io.micronaut.core.annotation.*;
```
Can you remove all changes that use `*` imports and revert to individual imports

graemerocher commented on inject/src/main/java/io/micronaut/context/AbstractBeanDefinition2.java:
```diff
@@ -0,0 +1,2338 @@
+        }
+    }
+
+    public static final class MethodReference extends MethodOrFieldReference {
```
Needs javadoc, also we already have an interface called `MethodReference` so needs to be named something else

graemerocher commented on core/src/main/java/io/micronaut/core/util/ImmutableSortedStringsArrayMap.java:
```diff
@@ -0,0 +1,540 @@
+@SuppressWarnings("ParameterNumber")
+@Internal
+@UsedByGeneratedCode
+public final class ImmutableSortedStringsArrayMap {
```
Make package private and have a factory method in `AnnotationUtil` create this

graemerocher commented on inject/src/main/java/io/micronaut/inject/annotation/AnnotationMetadataWriter.java:
```diff
@@ -450,19 +530,21 @@ private ClassWriter generateClassBytes() {
         GeneratorAdapter constructor = startConstructor(classWriter);
         DefaultAnnotationMetadata annotationMetadata = (DefaultAnnotationMetadata) this.annotationMetadata;
 
+        Map<String, Integer> defaultsStorage = new HashMap<>(3);
```
Can you explain what this `defaultsStorage` is used for?

dstepanov replied:
It's for sharing annotation defaults instead of calling it multiple times in the generated code.

graemerocher commented on inject/src/main/java/io/micronaut/context/AbstractExecutableMethodsDefinition.java:
```diff
@@ -0,0 +1,381 @@
+     * @return true if matches
+     */
+    @UsedByGeneratedCode
+    protected final boolean methodAtIndexMatches(int index, String name, Class[] argumentTypes) {
```
Where/how is this used?

dstepanov replied:
That's part of the `getMethod` "string switch" like code to check if the method at index matches the requested name + arguments.

graemerocher reviewed (changes requested):
In general I think if we are going to introduce a new metadata format this is something that should happen very rarely and we have to make sure we get it right and fix any problems in the previous format. With that in mind I have added some questions and comments on what needs to be improved.

I think that we need to reduce the complexity in `AbstractBeanDefinition` and `AbstractBeanDefinition2` (maybe we need a better name) should aim for that. I have a feeling we should have a more specialized `AbstractConfigPropsBeanDefinition` that moves all the logic for `@EachProperty / @EachBean / @ConfigProp` to a child class and reduces the complexity necessary in `AbstractBeanDefinition` to support configuration properties.

We also need to use the build time knowledge better and dispatch directly to the appropriate method rather than having too much complexity in the generic `getBeanForConstructorArgument` methods etc.

dstepanov commented:
I think I agree with most of the suggestions to limit runtime evaluation. Not sure about `AbstractConfigPropsBeanDefinition` the idea is nice but if you check there is no so much of the code that is needed only for config-prop beans. We would also need a new config-specific class extending `AbstractParametrizedBeanDefinition2`.

One thing regarding `ParametrizedBeanFactory` I have noticed there is a code inside `DefaultBeanContext` that does resolve value for the arguments inside `doCreateBean`. Do you know why does it need to be there instead of in `AbstractParametrizedBeanDefinition`?

dstepanov commented:
Regarding the native image size, I have observed in the GraalVM dashboard:
- `io.micronaut.*` code size in M2: 5.259 MB vs PR: 3.558 MB
- heap size with PR: -400 kB I see big amount of `HashMap$Node` entries are gone
- native image binary is smaller by 2.115MB

dstepanov commented:
I have addressed all of the comments, moved a lot of the resolutions to the compile-time.

Regarding the speed improvements on JVM, with the first changes, I have seen around 400ms to 390ms drop in the startup time, I would expect more improvements with the final changes.

I have updated the summary with the changes included.

jameskleeh commented on inject-groovy/src/test/groovy/io/micronaut/inject/field/FieldInjectionSpec.groovy:
```diff
@@ -166,5 +178,31 @@ class FieldInjectionSpec extends Specification {
         }
     }
 
+    static class F {
```
Is this class used somewhere else? Doesn't seem to be used in this spec

dstepanov replied:
Good catch, I have corrected the test

jameskleeh commented on inject-groovy/src/test/groovy/io/micronaut/inject/field/FieldInjectionSpec.groovy:
```diff
@@ -66,8 +63,8 @@ class FieldInjectionSpec extends Specification {
         e.d == null
-        e.value == "Default greeting"
-        e.property == "Default greeting"
+        e.value == null
```
I think we should maintain this behavior if possible

dstepanov replied:
That’s only Groovy related, Java code is returning nulls.
Do we want to have those inconsistencies?

jameskleeh replied:
No all good then. Should make a note in the breaks.adoc about it though

dstepanov replied:
I have added the note to the breaking changes

dstepanov committed:
Correct test and add breaking changes

jameskleeh commented:
@dstepanov Have you done a smoke test to ensure classes compiled with the old abstract classes still function correctly with the changes in this PR?

dstepanov commented:
@jameskleeh I didn't, but I have checked the changes again I don't see any runtime changes that can break it.

graemerocher commented:
@jameskleeh I will verify that

graemerocher commented on core/src/main/java/io/micronaut/core/util/ArrayUtils.java:
```diff
@@ -172,6 +175,18 @@ public static String toString(String delimiter, @Nullable Object[] array) {
+     * @param <T> The type of the array
+     * @return The array
+     */
+    public static <T> T[] toArray(Collection<T> collection, Class<T> arrayItemClass) {
```
Add `@UsedByGeneratedCode`

dstepanov replied:
Actually, I end up using generated `.toArray` instead of this method. I can delete it or I will just add `@since`

graemerocher commented on core/src/main/java/io/micronaut/core/util/ArrayUtils.java:
```diff
@@ -172,6 +175,18 @@ public static String toString(String delimiter, @Nullable Object[] array) {
+     * @param collection The collection
+     * @param arrayItemClass The array item class
+     * @param <T> The type of the array
+     * @return The array
```
Add `@since`

graemerocher commented on inject/src/main/java/io/micronaut/context/AbstractBeanDefinition2.java:
```diff
@@ -0,0 +1,1846 @@
+ * @since 3.0
+ */
+@Internal
+public class AbstractBeanDefinition2<T> extends AbstractBeanContextConditional implements BeanDefinition<T>, EnvironmentConfigurable {
```
Let's name this `AbstractInitializableBeanDefinition` since it is more descriptive of what it does vs the previous one

graemerocher commented on inject/src/main/java/io/micronaut/context/AbstractBeanDefinition2.java:
```diff
@@ -0,0 +1,1846 @@
+
+    private final Class<T> type;
+    private final AnnotationMetadata annotationMetadata;
+    private final Optional<String> scope;
```
Can we just store a nullable string here? It would be more consistent with the other parameters that are annotated with `@Nullable`

graemerocher commented on inject/src/main/java/io/micronaut/context/AbstractBeanDefinition2.java:
```diff
@@ -0,0 +1,1846 @@
+    @Nullable
+    private Environment environment;
+    private Set<Class<?>> exposedTypes;
+    private Optional<Argument<?>> containerElement;
```
Same here do we need to store the optoinal? Would it not be sufficient to just make this Nullable?

graemerocher reviewed (changes requested):
This is amazing work @dstepanov well done! I tested it mixing with other old byte code and there is no problem as far as I can tell also in JIT mode it reduces RSS memory by between 30-40MB and the JAR size is over 2MB smaller

I have a few small changes to request but then we should merge this. Good job!

dstepanov committed:
Address comments, change to generate exposed types at the runtime and pass the check if the type is container
